scio-extra/src/main/scala/com/spotify/scio/extra/voyager/syntax/SCollectionSyntax.scala

```diff
@@ -71,40 +81,41 @@ class VoyagerPairSCollectionOps(
    *
    * @param uri
    *   The [[VoyagerUri]].
-   * @param spaceType
-   *   The measurement for computing distance between entities. One of Euclidean, Cosine or Dot
-   *   (inner product).
+   * @param space
+   *   The space type to use when storing and comparing vectors.
+   * @param numDimensions
+   *   The number of dimensions per vector.
+   * @param indexM
+   *   Controls the degree of interconnectedness between vectors.
+   * @param efConstruction
+   *   Controls index quality, affecting the speed of addItem calls. Does not affect memory usage or
+   *   size of the index.
+   * @param randomSeed
+   *   A random seed to use when initializing the index's internal data structures.
    * @param storageDataType
-   *   The storage data type of the vectors at rest. One of Float8, Float32 or E4M3.
-   * @param dim
-   *   Number of dimensions in vectors.
-   * @param ef
-   *   The size of the dynamic list of neighbors used during construction time. This parameter
-   *   controls query time/accuracy tradeoff. More information can be found in the hnswlib
-   *   documentation https://github.com/nmslib/hnswlib.
-   * @param m
-   *   The number of outgoing connections in the graph.
+   *   The datatype to use under-the-hood when storing vectors.
    * @return
    *   A [[VoyagerUri]] representing where the index was written to.
    */
   @experimental
   def asVoyager(
     uri: VoyagerUri,
-    spaceType: SpaceType,
-    storageDataType: StorageDataType,
-    dim: Int,
-    ef: Long,
-    m: Long
+    space: SpaceType,
+    numDimensions: Int,
+    indexM: Long,
+    efConstruction: Long,
+    randomSeed: Long,
+    storageDataType: StorageDataType
   ): SCollection[VoyagerUri] = {
     implicit val remoteFileUtil: RemoteFileUtil = RemoteFileUtil.create(self.context.options)
     require(!uri.exists, s"Voyager URI ${uri.value} already exists")
 
     self.transform { in =>
-      val vectors = in.asIterableSideInput
-      self.context
-        .parallelize(Seq((): Unit))
-        .withSideInputs(vectors)
-        .map { case (_, ctx) =>
+      val count = in.count.asSingletonSideInput
+      in
+        .groupBy(_ => ()) // asIterableSideInput fails for large indexes
```

Comment:
It looks like `asIterableSideInput` has very poor performance when the collection can't fit in memory.
When the step indicates `Bytes written 9.82 GB`, The `Bytes read` from side input was going above `20GB` though the input should have been materialized only once.
Change implementation to group by single unique key. Add count side input to properly dimension the voyager index and avoid resizing.